src/container/intrusive_list.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <stdexcept>
+#include <yaclib/container/intrusive_node.hpp>
+
+namespace yaclib::container {
+
+template <typename T>
+class IntrusiveList {
+ public:
+  void PushBack(detail::IntrusiveNode<T>* node) noexcept {
+    node->LinkBefore(&_head);
+  }
+
+  void PushFront(detail::IntrusiveNode<T>* node) noexcept {
+    node->LinkBefore(_head._next);
+  }
+
+  // Returns nullptr if empty
+  T* PopFront() noexcept {
+    if (IsEmpty()) {
+      return nullptr;
+    }
+    auto* front = _head._next;
+    front->Unlink();
+    return front->AsItem();
+  }
+
+  // Returns nullptr if empty
+  T* PopBack() noexcept {
+    if (IsEmpty()) {
+      return nullptr;
+    }
+    auto* back = _head._prev;
+    back->Unlink();
+    return back->AsItem();
+  }
+
+  // Append (= move, re-link) all nodes from `that` list to the end of this list
+  // Post-condition: that.IsEmpty() == true
+  void Append(IntrusiveList& that) noexcept {
```

Comment:
- [ ] Remove `void Append(IntrusiveList& that) noexcept` method

We use copy/move ctor/assign or methods with the same semantics. Don't write both semantics in the same class